### PR TITLE
fix  Mongo storage error with GPTTreeIndex

### PR DESCRIPTION
**Pull request by aidenchen01**
when setup mongo storage for GPTTreeIndex,  it proodues error  like 
```
Traceback (most recent call last):
  File "/tmp/pycharm_project_19/test.py", line 94, in <module>
    GPTTreeIndex(nodes,storage_context=storage_context, service_context=service_context)
  File "/tmp/pycharm_project_19/llama_index/indices/tree/base.py", line 77, in __init__
    super().__init__(
  File "/tmp/pycharm_project_19/llama_index/indices/base.py", line 68, in __init__
    self._storage_context.index_store.add_index_struct(self._index_struct)
  File "/tmp/pycharm_project_19/llama_index/storage/index_store/keyval_index_store.py", line 37, in add_index_struct
    self._kvstore.put(key, data, collection=self._collection)
  File "/tmp/pycharm_project_19/llama_index/storage/kvstore/mongodb_kvstore.py", line 112, in put
    self._db[collection].replace_one(
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/collection.py", line 928, in replace_one
    self._update_retryable(
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/collection.py", line 836, in _update_retryable
    return self.__database.client._retryable_write(
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/mongo_client.py", line 1476, in _retryable_write
    return self._retry_with_session(retryable, func, s, None)
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/mongo_client.py", line 1349, in _retry_with_session
    return self._retry_internal(retryable, func, session, bulk)
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/_csot.py", line 105, in csot_wrapper
    return func(self, *args, **kwargs)
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/mongo_client.py", line 1390, in _retry_internal
    return func(session, sock_info, retryable)
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/collection.py", line 817, in _update
    return self._update(
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/collection.py", line 773, in _update
    result = sock_info.command(
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/pool.py", line 795, in command
    self._raise_connection_failure(error)
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/pool.py", line 767, in command
    return command(
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/network.py", line 127, in command
    request_id, msg, size, max_doc_size = message._op_msg(
  File "/home/username/.local/lib/python3.10/site-packages/pymongo/message.py", line 692, in _op_msg
    return _op_msg_uncompressed(flags, command, identifier, docs, opts)
bson.errors.InvalidDocument: documents must have only string keys, key was 0

```

In  the test.py before, I setup  storage context like 

```python
MongoDocumentStore.from_uri(uri="mongodb+srv://useer:password@something.mongodb.net/?retryWrites=true&w=majority")
```

and use GPTTreeIndex like:

```python
GPTTreeIndex(nodes,storage_context=storage_context, service_context=service_context)
```

### CODE REVIEW

**Comment by Disiok** [comment]
Thanks for the fix! This makes sense, just want to figure out minimize disruption to users who already have index struct saved with the dict format.

File: llama_index/storage/index_store/utils.py [hunk 1]
```diff
@@ -15,4 +15,4 @@ def json_to_index_struct(struct_dict: dict) -> IndexStruct:
     type = struct_dict[TYPE_KEY]
     data_dict = struct_dict[DATA_KEY]
     cls = INDEX_STRUCT_TYPE_TO_INDEX_STRUCT_CLASS[type]
-    return cls.from_dict(data_dict)
+    return cls.from_json(data_dict)
```
Comment (Disiok) [inline llama_index/storage/index_store/utils.py:18]: Could we try/catch `TypeError`, and fallback to loading the data as dict (to remain compatible with users who saved index in the past).

Comment (Disiok) [reply llama_index/storage/index_store/utils.py:18]: And add a note saying data blobs were stored as dict, but now is serialized to string first.

**Review by jerryjliu — APPROVED** [review]
sorry for the delay!

**Pull request merged**